aws/resource_aws_codepipeline_webhook.go

```diff
@@ -0,0 +1,336 @@
+				}, false),
+			},
+			"authentication_configuration": {
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				MinItems: 1,
+				Required: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"secret_token": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"allowed_ip_range": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validation.CIDRNetwork(0, 32),
+						},
+					},
+				},
+			},
+			"filter": {
+				Type:     schema.TypeSet,
+				ForceNew: true,
+				Required: true,
+				MinItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"json_path": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"match_equals": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"target_action": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"target_pipeline": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+		},
+	}
+}
+
+func extractCodePipelineWebhookAttr(d *schema.ResourceData, attr string) (map[string]interface{}, error) {
+	if v, ok := d.GetOk(attr); ok {
+		l := v.([]interface{})
+		if len(l) <= 0 {
+			return nil, fmt.Errorf("Attribute %s is missing", attr)
+		}
+
+		data := l[0].(map[string]interface{})
+		return data, nil
+	}
+
+	return nil, fmt.Errorf("Could not find attribute %s", attr)
+}
+
+func extractCodePipelineWebhookRules(filters *schema.Set) ([]*codepipeline.WebhookFilterRule, error) {
+	var rules []*codepipeline.WebhookFilterRule
+
+	for _, f := range filters.List() {
+		r := f.(map[string]interface{})
+		filter := codepipeline.WebhookFilterRule{
+			JsonPath:    aws.String(r["json_path"].(string)),
+			MatchEquals: aws.String(r["match_equals"].(string)),
+		}
+
+		rules = append(rules, &filter)
+	}
+
+	return rules, nil
+}
+
+func extractCodePipelineWebhookAuthConfig(authType string, authConfig map[string]interface{}) (*codepipeline.WebhookAuthConfiguration, error) {
+	var conf codepipeline.WebhookAuthConfiguration
+	switch authType {
+	case codepipeline.WebhookAuthenticationTypeIp:
+		ipRange := authConfig["allowed_ip_range"].(string)
+		if ipRange == "" {
+			return nil, fmt.Errorf("An IP range must be set when using IP-based auth")
+		}
+
+		conf.AllowedIPRange = &ipRange
+
+		break
+	case codepipeline.WebhookAuthenticationTypeGithubHmac:
+		secretToken := authConfig["secret_token"].(string)
+		if secretToken == "" {
+			return nil, fmt.Errorf("Secret token must be set when using GITHUB_HMAC")
+		}
+
+		conf.SecretToken = &secretToken
+		break
+	case codepipeline.WebhookAuthenticationTypeUnauthenticated:
+		break
+	default:
```

Comment:
The schema handling for the `auth_type` argument prevents this case from ever getting hit.